### DISCUSSION TITLE
After Training and Inference, some nodes are not predicted

### DISCUSSION THREAD

**Discussion by Jaspermoray** [talmolab/sleap#1164] [Help!]
Hi all,

I'm trying to use SLEAP to track crayfish, but I've run into a problem: After I choose initial instances, I run a top down training on 20 random frames, and only SOME of the nodes in those frames are predicted. When I return to the video scrolling part of the program, the predicted frames have less than half of the nodes, with the others just greyed out. 
    
If it helps, I noticed that when I turned on visualization of the training, the frame appears to be cut off at the edges, and those nodes that are cut off are not predicted. Does anybody know why this would be?

Best,
Jasper Coles Hood

**Comment by roomrys** [comment]
Hi @Jaspermoray,

It sounds like the crop size is for your top-down model is too small. Can you ensure that the crop size is set to Auto for both model configurations (in the Training Pipeline GUI)?

![image](https://user-images.githubusercontent.com/38435167/216668251-84d392aa-668f-4d77-83c1-26d8acb2b445.png)

Let us know if that helps.

Thanks,
Liezl

**Reply by Jaspermoray** [reply]
Yeah, It is set to auto. I could try increasing it I suppose

**Reply by roomrys** [reply]
Hi @Jaspermoray,

Following-up... Just wondering if your subject(s) take(s) up most of the frame? If so, bottom-up may be a better approach than top-down. Most of the benefit of top-down is seen when the subjects are relatively small to the entire field of view.

Let me know if you are having any success. If none of these suggestions seem to be working, I can also have a look if you don't mind sharing your dataset with [lmaree@salk.edu](mailto:lmaree@salk.edu). We'll follow up here for posterity of course.

Thanks,
Liezl

**Reply by Jaspermoray** [reply]
Oh, okay. Thats interesting. I'll give that another shot. To be honest, it was my first look at the software, so I just uploaded a clear video of my hand and wiggled my fingers one by one, and it was cutting out my fingers. But I'll definitely give that a shot.

**Reply by Jaspermoray** [reply]
Unfortunately, it gives the following error: "Cannot run the bottom up pipeline because the skeleton is not an arborescence. There are multiple root nodes..... There are nodes which are target in multiple edges." Sounds like bottom up only works for simplified skeletons?

**Reply by roomrys** [reply]
Hi @Jaspermoray,

Sorry, I should have mentioned that the [skeleton](https://sleap.ai/guides/skeletons.html?highlight=skeleton) needs to branch out like a tree ([tree-graph](https://calcworkshop.com/trees-graphs/tree-graph/)) for bottom-up. 
![image](https://user-images.githubusercontent.com/38435167/217105026-b396275b-3e9d-4e56-9311-0e4804f02ec2.png)
To be a tree, each node in the skeleton can only have one parent (or one source). So when you are connecting the edges of your skeleton, you should ensure that each destination node only has one source (although any source can have multiple destinations - checkout node b above which is a source to multiple destinations: d, e, and f).

Let me know if this helps!

Thanks,
Liezl

**Reply by Jaspermoray** [reply]
Hi Leizl,

 Thank you for your help, I really appreciate this. Does the source node have to be labeled specifically? In the picture below, Supposedly the wrist node would be the source, but it doesn't recognize that...
<img width="620" alt="Screen Shot 2023-02-07 at 11 14 15 AM" src="https://user-images.githubusercontent.com/94078678/217300573-cd048637-9856-4c45-a292-a7f191dad83f.png">

Best,
Jasper

**Reply by roomrys** [reply]
Hi @Jaspermoray,

When creating edges in SLEAP, all source nodes will be listed in the "Source" column and will also be the first node selected via the drop-down boxes, i.e. in the image below, PointerBase is the source and PointerEnd is the destination (source to destination).
![image](https://user-images.githubusercontent.com/38435167/217335481-797048a9-62cc-4b81-9288-7d874bfe2654.png)

Thanks,
Liezl
